diff -urNp coreutils-8.4-orig/src/stat.c coreutils-8.4/src/stat.c
--- coreutils-8.4-orig/src/stat.c	2013-07-24 17:12:38.506951106 +0200
+++ coreutils-8.4/src/stat.c	2013-07-24 17:13:44.054031351 +0200
@@ -693,7 +693,6 @@ print_stat (char *pformat, size_t prefix
       out_uint (pformat, prefix_len, statbuf->st_uid);
       break;
     case 'U':
-      setpwent ();
       pw_ent = getpwuid (statbuf->st_uid);
       out_string (pformat, prefix_len,
                   pw_ent ? pw_ent->pw_name : "UNKNOWN");
@@ -702,7 +701,6 @@ print_stat (char *pformat, size_t prefix
       out_uint (pformat, prefix_len, statbuf->st_gid);
       break;
     case 'G':
-      setgrent ();
       gw_ent = getgrgid (statbuf->st_gid);
       out_string (pformat, prefix_len,
                   gw_ent ? gw_ent->gr_name : "UNKNOWN");
