*** ../binutils-2.23.52.0.1.orig/ld/emulparams/aarch64linux.sh	2013-04-09 17:12:43.108529894 +0100
--- ld/emulparams/aarch64linux.sh	2013-04-09 17:15:04.382533810 +0100
*************** OTHER_SECTIONS='.note.gnu.arm.ident 0 :
*** 34,36 ****
--- 34,46 ----
  ATTRS_SECTIONS='.ARM.attributes 0 : { KEEP (*(.ARM.attributes)) KEEP (*(.gnu.attributes)) }'
  # Ensure each PLT entry is aligned to a cache line.
  PLT=".plt          ${RELOCATING-0} : ALIGN(16) { *(.plt)${IREL_IN_PLT+ *(.iplt)} }"
+ 
+ # Linux modifies the default library search path to first include
+ # a 64-bit specific directory.
+ case "$target" in
+   aarch64*-linux*)
+     case "$EMULATION_NAME" in
+       aarch64linux*) LIBPATH_SUFFIX=64 ;;
+     esac
+     ;;
+ esac
