*** binutils-2.20.51.0.2.orig/bfd/elfcode.h	2012-04-05 09:58:31.595825648 +0100
--- binutils-2.20.51.0.2/bfd/elfcode.h	2012-04-05 09:58:59.042824782 +0100
*************** elf_checksum_contents (bfd *abfd,
*** 1186,1194 ****
        elf_swap_shdr_out (abfd, &i_shdr, &x_shdr);
        (*process) (&x_shdr, sizeof x_shdr, arg);
  
        if (i_shdr.contents)
  	(*process) (i_shdr.contents, i_shdr.sh_size, arg);
!       else
  	{
  	  asection *sec;
  
--- 1186,1196 ----
        elf_swap_shdr_out (abfd, &i_shdr, &x_shdr);
        (*process) (&x_shdr, sizeof x_shdr, arg);
  
+       /* PR ld/12451:
+ 	 Process the section's contents, if it has some.  Read them in if necessary.  */
        if (i_shdr.contents)
  	(*process) (i_shdr.contents, i_shdr.sh_size, arg);
!       else if (i_shdr.sh_type != SHT_NOBITS)
  	{
  	  asection *sec;
  
