--- ./util.c	2006-12-11 19:54:39.000000000 +0100
+++ ./util.c	2013-05-03 13:50:02.756377289 +0200
@@ -150,11 +150,30 @@
      voidp buf;
      unsigned int cnt;
 {
+  int len;
 #ifdef SSIZE_MAX
   if (SSIZE_MAX < cnt)
     cnt = SSIZE_MAX;
 #endif
-  return read (fd, buf, cnt);
+  len = read (fd, buf, cnt);
+
+#if defined F_SETFL && O_NONBLOCK && defined EAGAIN
+  /* Input files are opened O_NONBLOCK for security reasons.  On some
+     file systems this can cause read to fail with errno == EAGAIN.  */
+  if (len < 0 && errno == EAGAIN)
+    {
+      int flags = fcntl (fd, F_GETFL);
+      if (0 <= flags)
+        {
+          if (! (flags & O_NONBLOCK))
+            errno = EAGAIN;
+          else if (fcntl (fd, F_SETFL, flags & ~O_NONBLOCK) != -1)
+            len = read (fd, buf, cnt);
+        }
+    }
+#endif
+
+  return len;
 }
 
 /* Likewise for 'write'.  */
